File: XCTestBootstrap/Configuration/FBXCTestConfiguration.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBXCTestConfiguration.h"

#import <FBControlCore/FBControlCore.h>

#import "FBXCTestConfiguration.h"
#import "FBCodeCoverageConfiguration.h"
#import "FBXCTestProcess.h"
#import "XCTestBootstrapError.h"

FBXCTestType const FBXCTestTypeApplicationTest = FBXCTestTypeApplicationTestValue;
FBXCTestType const FBXCTestTypeLogicTest = @"logic-test";
FBXCTestType const FBXCTestTypeListTest = @"list-test";
FBXCTestType const FBXCTestTypeUITest = @"ui-test";

@implementation FBXCTestConfiguration

#pragma mark Initializers

- (instancetype)initWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _processUnderTestEnvironment = environment ?: @{};
  _workingDirectory = workingDirectory;
  _testBundlePath = testBundlePath;
  _waitForDebugger = waitForDebugger;

  NSString *timeoutFromEnv = NSProcessInfo.processInfo.environment[@"FB_TEST_TIMEOUT"];
  if (timeoutFromEnv) {
    _testTimeout = timeoutFromEnv.intValue;
  } else {
    _testTimeout = timeout > 0 ? timeout : [self defaultTimeout];
  }

  return self;
}
#pragma mark Public


- (NSTimeInterval)defaultTimeout
{
  // TSAN is known to slow down all tests from running, bump the default test timeout to 1800s (instead of 500s).
#if __has_feature(thread_sanitizer) || defined(__SANITIZE_THREAD__)
  return 1800.0;
#endif
  return 500;
}

- (NSString *)testType
{
  NSAssert(NO, @"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self.class), NSStringFromSelector(_cmd));
  return nil;
}

- (NSDictionary<NSString *, NSString *> *)buildEnvironmentWithEntries:(NSDictionary<NSString *, NSString *> *)entries
{
  NSMutableDictionary<NSString *, NSString *> *parentEnvironment = NSProcessInfo.processInfo.environment.mutableCopy;
  [parentEnvironment removeObjectsForKeys:@[
    @"XCTestConfigurationFilePath",
  ]];

  NSMutableDictionary<NSString *, NSString *> *environmentOverrides = [NSMutableDictionary dictionary];
  NSString *xctoolTestEnvPrefix = @"XCTOOL_TEST_ENV_";
  for (NSString *key in parentEnvironment) {
    if ([key hasPrefix:xctoolTestEnvPrefix]) {
      NSString *childKey = [key substringFromIndex:xctoolTestEnvPrefix.length];
      environmentOverrides[childKey] = parentEnvironment[key];
    }
  }
  [environmentOverrides addEntriesFromDictionary:entries];
  NSMutableDictionary<NSString *, NSString *> *environment = parentEnvironment.mutableCopy;
  for (NSString *key in environmentOverrides) {
    NSString *childKey = key;
    environment[childKey] = environmentOverrides[key];
  }
  return environment.copy;
}

#pragma mark NSObject

- (NSString *)description
{
  NSData *data = [NSJSONSerialization dataWithJSONObject:self.jsonSerializableRepresentation options:0 error:NULL];
  NSParameterAssert(data);
  return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

- (BOOL)isEqual:(FBXCTestConfiguration *)object
{
  // Class must match exactly in the class-cluster
  if (![object isMemberOfClass:self.class]) {
    return NO;
  }
  return (self.processUnderTestEnvironment == object.processUnderTestEnvironment || [self.processUnderTestEnvironment isEqualToDictionary:object.processUnderTestEnvironment])
      && (self.workingDirectory == object.workingDirectory || [self.workingDirectory isEqualToString:object.workingDirectory])
      && (self.testBundlePath == object.testBundlePath || [self.testBundlePath isEqualToString:object.testBundlePath])
      && (self.testType == object.testType || [self.testType isEqualToString:object.testType])
      && (self.waitForDebugger == object.waitForDebugger)
      && (self.testTimeout == object.testTimeout);
}

- (NSUInteger)hash
{
  return self.processUnderTestEnvironment.hash ^ self.workingDirectory.hash ^ self.testBundlePath.hash ^ self.testType.hash ^ ((NSUInteger) self.waitForDebugger) ^ ((NSUInteger) self.testTimeout);
}

#pragma mark JSON

static NSString *const KeyEnvironment = @"environment";
static NSString *const KeyListTestsOnly = @"list_only";
static NSString *const KeyOSLogPath = @"os_log_path";
static NSString *const KeyRunnerAppPath = @"test_host_path";
static NSString *const KeyRunnerTargetPath = @"test_target_path";
static NSString *const KeyTestArtifactsFilenameGlobs = @"test_artifacts_filename_globs";
static NSString *const KeyTestBundlePath = @"test_bundle_path";
static NSString *const KeyTestFilter = @"test_filter";
static NSString *const KeyTestMirror = @"test_mirror";
static NSString *const KeyTestTimeout = @"test_timeout";
static NSString *const KeyTestType = @"test_type";
static NSString *const KeyVideoRecordingPath = @"video_recording_path";
static NSString *const KeyWaitForDebugger = @"wait_for_debugger";
static NSString *const KeyWorkingDirectory = @"working_directory";

- (id)jsonSerializableRepresentation
{
  return @{
    KeyEnvironment: self.processUnderTestEnvironment,
    KeyWorkingDirectory: self.workingDirectory,
    KeyTestBundlePath: self.testBundlePath,
    KeyTestType: self.testType,
    KeyListTestsOnly: @NO,
    KeyWaitForDebugger: @(self.waitForDebugger),
    KeyTestTimeout: @(self.testTimeout),
  };
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  return self;
}

@end

@implementation FBListTestConfiguration

#pragma mark Initializers

+ (instancetype)configurationWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath runnerAppPath:(nullable NSString *)runnerAppPath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout architectures:(nonnull NSSet<NSString *> *)architectures
{
  return [[FBListTestConfiguration alloc] initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath runnerAppPath:runnerAppPath waitForDebugger:waitForDebugger timeout:timeout architectures:architectures];
}

- (instancetype)initWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath runnerAppPath:(nullable NSString *)runnerAppPath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout architectures:(nonnull NSSet<NSString *> *)architectures
{
  self = [super initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath waitForDebugger:waitForDebugger timeout:timeout];
  if (!self) {
    return nil;
  }

  _runnerAppPath = runnerAppPath;
  _architectures = architectures;

  return self;
}

#pragma mark Public

- (NSString *)testType
{
  return FBXCTestTypeListTest;
}

#pragma mark JSON

- (id)jsonSerializableRepresentation
{
  NSMutableDictionary<NSString *, id> *json = [NSMutableDictionary dictionaryWithDictionary:[super jsonSerializableRepresentation]];
  json[KeyListTestsOnly] = @YES;
  json[KeyRunnerAppPath] = _runnerAppPath ?: NSNull.null;
  return [json copy];
}

@end

@implementation FBTestManagerTestConfiguration

#pragma mark Initializers

+ (instancetype)configurationWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout runnerAppPath:(NSString *)runnerAppPath testTargetAppPath:(NSString *)testTargetAppPath testFilter:(NSString *)testFilter videoRecordingPath:(NSString *)videoRecordingPath testArtifactsFilenameGlobs:(nullable NSArray<NSString *> *)testArtifactsFilenameGlobs osLogPath:(nullable NSString *)osLogPath
{
  return [[FBTestManagerTestConfiguration alloc] initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath waitForDebugger:waitForDebugger timeout:timeout runnerAppPath:runnerAppPath testTargetAppPath:testTargetAppPath testFilter:testFilter videoRecordingPath:videoRecordingPath testArtifactsFilenameGlobs:testArtifactsFilenameGlobs osLogPath:osLogPath];
}

- (instancetype)initWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout runnerAppPath:(NSString *)runnerAppPath testTargetAppPath:(NSString *)testTargetAppPath testFilter:(NSString *)testFilter videoRecordingPath:(NSString *)videoRecordingPath testArtifactsFilenameGlobs:(NSArray<NSString *> *)testArtifactsFilenameGlobs osLogPath:(nullable NSString *)osLogPath
{
  self = [super initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath waitForDebugger:waitForDebugger timeout:timeout];
  if (!self) {
    return nil;
  }

  _runnerAppPath = runnerAppPath;
  _testTargetAppPath = testTargetAppPath;
  _testFilter = testFilter;
  _videoRecordingPath = videoRecordingPath;
  _testArtifactsFilenameGlobs = testArtifactsFilenameGlobs;
  _osLogPath = osLogPath;

  return self;
}

#pragma mark Public

- (NSString *)testType
{
  return _testTargetAppPath != nil ? FBXCTestTypeUITest : FBXCTestTypeApplicationTest;
}

#pragma mark JSON

- (id)jsonSerializableRepresentation
{
  NSMutableDictionary<NSString *, id> *json = [NSMutableDictionary dictionaryWithDictionary:[super jsonSerializableRepresentation]];
  json[KeyRunnerAppPath] = self.runnerAppPath;
  json[KeyRunnerTargetPath] = self.testTargetAppPath;
  json[KeyTestFilter] = self.testFilter ?: NSNull.null;
  json[KeyVideoRecordingPath] = self.videoRecordingPath ?: NSNull.null;
  json[KeyTestArtifactsFilenameGlobs] = self.testArtifactsFilenameGlobs ?: NSNull.null;
  json[KeyOSLogPath] =  self.osLogPath ?: NSNull.null;
  return [json copy];
}

@end

@implementation FBLogicTestConfiguration

#pragma mark Initializers

+ (instancetype)configurationWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout testFilter:(NSString *)testFilter mirroring:(FBLogicTestMirrorLogs)mirroring coverageConfiguration:(nullable FBCodeCoverageConfiguration *)coverageConfiguration binaryPath:(nullable NSString *)binaryPath logDirectoryPath:(NSString *)logDirectoryPath architectures:(nonnull NSSet<NSString *> *)architectures
{
  return [[FBLogicTestConfiguration alloc] initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath waitForDebugger:waitForDebugger timeout:timeout testFilter:testFilter  mirroring:mirroring coverageConfiguration:coverageConfiguration binaryPath:binaryPath logDirectoryPath:logDirectoryPath architectures:architectures];
}

- (instancetype)initWithEnvironment:(NSDictionary<NSString *, NSString *> *)environment workingDirectory:(NSString *)workingDirectory testBundlePath:(NSString *)testBundlePath waitForDebugger:(BOOL)waitForDebugger timeout:(NSTimeInterval)timeout testFilter:(NSString *)testFilter mirroring:(FBLogicTestMirrorLogs)mirroring coverageConfiguration:(nullable FBCodeCoverageConfiguration *)coverageConfiguration binaryPath:(nullable NSString *)binaryPath logDirectoryPath:(NSString *)logDirectoryPath architectures:(nonnull NSSet<NSString *> *)architectures
{
  self = [super initWithEnvironment:environment workingDirectory:workingDirectory testBundlePath:testBundlePath waitForDebugger:waitForDebugger timeout:timeout];
  if (!self) {
    return nil;
  }

  _testFilter = testFilter;
  _mirroring = mirroring;
  _coverageConfiguration = coverageConfiguration;
  _binaryPath = binaryPath;
  _logDirectoryPath = logDirectoryPath;
  _architectures = architectures;

  return self;
}

#pragma mark Public

- (NSString *)testType
{
  return FBXCTestTypeLogicTest;
}

#pragma mark JSON

- (id)jsonSerializableRepresentation
{
  NSMutableDictionary<NSString *, id> *json = [NSMutableDictionary dictionaryWithDictionary:[super jsonSerializableRepresentation]];
  json[KeyTestFilter] = self.testFilter ?: NSNull.null;
  return [json copy];
}

@end

File: XCTestBootstrap/Configuration/FBXCTestConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

// MARK: JSON Keys

private let KeyEnvironment = "environment"
private let KeyListTestsOnly = "list_only"
private let KeyOSLogPath = "os_log_path"
private let KeyRunnerAppPath = "test_host_path"
private let KeyRunnerTargetPath = "test_target_path"
private let KeyTestArtifactsFilenameGlobs = "test_artifacts_filename_globs"
private let KeyTestBundlePath = "test_bundle_path"
private let KeyTestFilter = "test_filter"
private let KeyTestMirror = "test_mirror"
private let KeyTestTimeout = "test_timeout"
private let KeyTestType = "test_type"
private let KeyVideoRecordingPath = "video_recording_path"
private let KeyWaitForDebugger = "wait_for_debugger"
private let KeyWorkingDirectory = "working_directory"

// MARK: Defaults

private let kDefaultTimeoutValue: TimeInterval = 500

// MARK: - Test Types

/// The type of an xctest execution, as reported in ocunit-shim events.
public struct FBXCTestType: RawRepresentable, Hashable, Sendable {
  public let rawValue: String

  public init(rawValue: String) {
    self.rawValue = rawValue
  }

  public static let applicationTest = FBXCTestType(rawValue: "application-test")
  public static let logicTest = FBXCTestType(rawValue: "logic-test")
  public static let listTest = FBXCTestType(rawValue: "list-test")
  public static let uiTest = FBXCTestType(rawValue: "ui-test")
}

/// Where a logic test's output is mirrored to, in addition to the reporter.
public struct FBLogicTestMirrorLogs: OptionSet, Sendable {
  public let rawValue: UInt

  public init(rawValue: UInt) {
    self.rawValue = rawValue
  }

  public static let fileLogs = FBLogicTestMirrorLogs(rawValue: 1 << 0)
  public static let logger = FBLogicTestMirrorLogs(rawValue: 1 << 1)
}

// MARK: - FBXCTestConfiguration

public class FBXCTestConfiguration: NSObject, NSCopying {

  public let processUnderTestEnvironment: [String: String]
  public let workingDirectory: String
  public let testBundlePath: String
  public let waitForDebugger: Bool
  public let testTimeout: TimeInterval

  var testType: FBXCTestType {
    fatalError("-[\(type(of: self)) testType] is abstract and should be overridden")
  }

  public init(environment: [String: String], workingDirectory: String, testBundlePath: String, waitForDebugger: Bool, timeout: TimeInterval) {
    self.processUnderTestEnvironment = environment
    self.workingDirectory = workingDirectory
    self.testBundlePath = testBundlePath
    self.waitForDebugger = waitForDebugger

    if let timeoutFromEnv = ProcessInfo.processInfo.environment["FB_TEST_TIMEOUT"],
      let envTimeout = TimeInterval(timeoutFromEnv)
    {
      self.testTimeout = envTimeout
    } else {
      self.testTimeout = timeout > 0 ? timeout : kDefaultTimeoutValue
    }
    super.init()
  }

  // MARK: NSObject

  public override var description: String {
    guard let data = try? JSONSerialization.data(withJSONObject: jsonSerializableRepresentation(), options: []) else {
      return super.description
    }
    return String(data: data, encoding: .utf8) ?? super.description
  }

  public override func isEqual(_ object: Any?) -> Bool {
    guard let other = object as? FBXCTestConfiguration else { return false }
    guard type(of: other) == type(of: self) else { return false }
    return processUnderTestEnvironment == other.processUnderTestEnvironment
      && workingDirectory == other.workingDirectory
      && testBundlePath == other.testBundlePath
      && testType.rawValue == other.testType.rawValue
      && waitForDebugger == other.waitForDebugger
      && testTimeout == other.testTimeout
  }

  public override var hash: Int {
    (processUnderTestEnvironment as NSDictionary).hash ^ (workingDirectory as NSString).hash ^ (testBundlePath as NSString).hash ^ (testType.rawValue as NSString).hash ^ (waitForDebugger ? 1 : 0) ^ Int(testTimeout)
  }

  // MARK: Public

  func buildEnvironment(withEntries entries: [String: String]) -> [String: String] {
    var parentEnvironment = ProcessInfo.processInfo.environment
    parentEnvironment.removeValue(forKey: "XCTestConfigurationFilePath")

    var environmentOverrides: [String: String] = [:]
    let xctoolTestEnvPrefix = "XCTOOL_TEST_ENV_"
    for (key, value) in parentEnvironment {
      if key.hasPrefix(xctoolTestEnvPrefix) {
        let childKey = String(key.dropFirst(xctoolTestEnvPrefix.count))
        environmentOverrides[childKey] = value
      }
    }
    for (key, value) in entries {
      environmentOverrides[key] = value
    }
    var environment = parentEnvironment
    for (key, value) in environmentOverrides {
      environment[key] = value
    }
    return environment
  }

  // MARK: JSON

  func jsonSerializableRepresentation() -> [String: Any] {
    [
      KeyEnvironment: processUnderTestEnvironment,
      KeyWorkingDirectory: workingDirectory,
      KeyTestBundlePath: testBundlePath,
      KeyTestType: testType.rawValue,
      KeyListTestsOnly: false,
      KeyWaitForDebugger: waitForDebugger,
      KeyTestTimeout: testTimeout,
    ]
  }

  // MARK: NSCopying

  public func copy(with zone: NSZone? = nil) -> Any {
    self
  }
}

// MARK: - FBListTestConfiguration

public final class FBListTestConfiguration: FBXCTestConfiguration {

  public let architectures: Set<String>
  public let runnerAppPath: String?

  public static func configuration(withEnvironment environment: [String: String], workingDirectory: String, testBundlePath: String, runnerAppPath: String?, waitForDebugger: Bool, timeout: TimeInterval, architectures: Set<String>) -> FBListTestConfiguration {
    FBListTestConfiguration(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, runnerAppPath: runnerAppPath, waitForDebugger: waitForDebugger, timeout: timeout, architectures: architectures)
  }

  public init(environment: [String: String], workingDirectory: String, testBundlePath: String, runnerAppPath: String?, waitForDebugger: Bool, timeout: TimeInterval, architectures: Set<String>) {
    self.runnerAppPath = runnerAppPath
    self.architectures = architectures
    super.init(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, waitForDebugger: waitForDebugger, timeout: timeout)
  }

  override var testType: FBXCTestType {
    FBXCTestType.listTest
  }

  override func jsonSerializableRepresentation() -> [String: Any] {
    var json = super.jsonSerializableRepresentation()
    json[KeyListTestsOnly] = true
    json[KeyRunnerAppPath] = runnerAppPath ?? NSNull()
    return json
  }
}

// MARK: - FBTestManagerTestConfiguration

public final class FBTestManagerTestConfiguration: FBXCTestConfiguration {

  public let runnerAppPath: String
  public let testTargetAppPath: String?
  public let testFilter: String?
  public let osLogPath: String?
  public let videoRecordingPath: String?
  public let testArtifactsFilenameGlobs: [String]?

  public static func configuration(withEnvironment environment: [String: String], workingDirectory: String, testBundlePath: String, waitForDebugger: Bool, timeout: TimeInterval, runnerAppPath: String, testTargetAppPath: String?, testFilter: String?, videoRecordingPath: String?, testArtifactsFilenameGlobs: [String]?, osLogPath: String?) -> FBTestManagerTestConfiguration {
    FBTestManagerTestConfiguration(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, waitForDebugger: waitForDebugger, timeout: timeout, runnerAppPath: runnerAppPath, testTargetAppPath: testTargetAppPath, testFilter: testFilter, videoRecordingPath: videoRecordingPath, testArtifactsFilenameGlobs: testArtifactsFilenameGlobs, osLogPath: osLogPath)
  }

  public init(environment: [String: String], workingDirectory: String, testBundlePath: String, waitForDebugger: Bool, timeout: TimeInterval, runnerAppPath: String, testTargetAppPath: String?, testFilter: String?, videoRecordingPath: String?, testArtifactsFilenameGlobs: [String]?, osLogPath: String?) {
    self.runnerAppPath = runnerAppPath
    self.testTargetAppPath = testTargetAppPath
    self.testFilter = testFilter
    self.videoRecordingPath = videoRecordingPath
    self.testArtifactsFilenameGlobs = testArtifactsFilenameGlobs
    self.osLogPath = osLogPath
    super.init(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, waitForDebugger: waitForDebugger, timeout: timeout)
  }

  override var testType: FBXCTestType {
    testTargetAppPath != nil ? FBXCTestType.uiTest : FBXCTestType.applicationTest
  }

  override func jsonSerializableRepresentation() -> [String: Any] {
    var json = super.jsonSerializableRepresentation()
    json[KeyRunnerAppPath] = runnerAppPath
    if let testTargetAppPath { json[KeyRunnerTargetPath] = testTargetAppPath }
    if let testFilter { json[KeyTestFilter] = testFilter }
    if let videoRecordingPath { json[KeyVideoRecordingPath] = videoRecordingPath }
    if let testArtifactsFilenameGlobs { json[KeyTestArtifactsFilenameGlobs] = testArtifactsFilenameGlobs }
    if let osLogPath { json[KeyOSLogPath] = osLogPath }
    return json
  }
}

// MARK: - FBLogicTestConfiguration

public final class FBLogicTestConfiguration: FBXCTestConfiguration {

  public let testFilter: String?
  public let mirroring: FBLogicTestMirrorLogs
  public let coverageConfiguration: FBCodeCoverageConfiguration?
  public let binaryPath: String?
  public let logDirectoryPath: String?
  public let architectures: Set<String>
  public let injectLibraries: [String]

  public static func configuration(withEnvironment environment: [String: String], workingDirectory: String, testBundlePath: String, waitForDebugger: Bool, timeout: TimeInterval, testFilter: String?, mirroring: FBLogicTestMirrorLogs, coverageConfiguration: FBCodeCoverageConfiguration?, binaryPath: String?, logDirectoryPath: String?, architectures: Set<String>) -> FBLogicTestConfiguration {
    FBLogicTestConfiguration(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, waitForDebugger: waitForDebugger, timeout: timeout, testFilter: testFilter, mirroring: mirroring, coverageConfiguration: coverageConfiguration, binaryPath: binaryPath, logDirectoryPath: logDirectoryPath, architectures: architectures)
  }

  public init(environment: [String: String], workingDirectory: String, testBundlePath: String, waitForDebugger: Bool, timeout: TimeInterval, testFilter: String?, mirroring: FBLogicTestMirrorLogs, coverageConfiguration: FBCodeCoverageConfiguration?, binaryPath: String?, logDirectoryPath: String?, architectures: Set<String>, injectLibraries: [String] = []) {
    self.testFilter = testFilter
    self.mirroring = mirroring
    self.coverageConfiguration = coverageConfiguration
    self.binaryPath = binaryPath
    self.logDirectoryPath = logDirectoryPath
    self.architectures = architectures
    self.injectLibraries = injectLibraries
    super.init(environment: environment, workingDirectory: workingDirectory, testBundlePath: testBundlePath, waitForDebugger: waitForDebugger, timeout: timeout)
  }

  override var testType: FBXCTestType {
    FBXCTestType.logicTest
  }

  override func jsonSerializableRepresentation() -> [String: Any] {
    var json = super.jsonSerializableRepresentation()
    json[KeyTestFilter] = testFilter ?? NSNull()
    return json
  }
}
